src/hotspot/share/opto/lcm.cpp

```diff
@@ -76,6 +76,34 @@ static bool needs_explicit_null_check_for_read(Node *val) {
   return true;
 }
 
+void PhaseCFG::move_into(Node* n, Block* b) {
+  Block* old = get_block_for_node(n);
+  old->find_remove(n);
+  b->add_inst(n);
+  map_node_to_block(n, b);
+  // Check for Mach projections that also need to be moved.
+  for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+    Node* out = n->fast_out(i);
+    if (!out->is_MachProj()) {
+      continue;
+    }
+    assert(!n->is_MachProj(), "nested projections are not allowed");
+    move_into(out, b);
+  }
+}
+
+void PhaseCFG::maybe_hoist_into(Node* n, Block* b) {
```

Comment:
Consider adding asserts into these 2 new methods to make sure that they operate only on **data** and not control nodes.

Reply:
Thanks, done (commit b198fca8).